### ISSUE TITLE
[Bug] Very slow performance when making calls to ASP.NET (ASMX) web service.

### ISSUE THREAD

**Issue by dmsi-leo** [xamarin/Xamarin.Forms#10960]
### Description
After upgrading our Xamarin Forms project from VS 2017 (Xamarin Forms 4.2.0.848062) to VS 2019 (Xamarin Forms 4.6.0.800) during stress and performance testing we noticed significant performance degradation when downloading files from ASP.NET (ASMX) web service. This is happening for Android and iOS but the problem is especially severe for Android.
For Android project we tried to set:
   HTTPClient Implementation build option to Android,
   SSL/TLS Implementation build option to Native TLS 1.2+
but it didn't improve performance by much.

Testing procedure:
Our Xamarin Forms app sequentially downloads 150 files ( each files size is roughly 1 MB) from ASP.NET (ASMX) web service.
The results of our testing:
Android:
App built with VS 2019 (Xamarin Forms 4.6.0.800) is **35-45%** slower then app built with VS 2017 (Xamarin Forms 4.2.0.848062).
iOS:
App built with VS 2019 (Xamarin Forms 4.6.0.800) is **10-20%** slower then app built with VS 2017 (Xamarin Forms 4.2.0.848062).

We tried different IIS Authentication methods (Basic, Digest, Windows) and web service binding protocols (HTTP and HTTPS) but that didn't improve performance.

This is very serious issue for us and it is important to improve performance at least to the level of VS 2017 (Xamarin Forms 4.2.0.848062) build.

Any help or suggestions or workarounds will be greatly appreciated.

Thanks for help,

Leo

### Steps to Reproduce

1. 
2. 
3. 

### Basic Information

- Version with issue: 4.6.0.800
- Last known good version: 4.2.0.848062
- IDE: Microsoft Visual Studio Enterprise 2019 Version 16.5.4
- Platform Target Frameworks: 
  - iOS:  10.3
  - Android: 10.0 
  - UWP:  18362 
- Android Support Library Version: 
- Nuget Packages: Xamarin.Essentials, SyncFusion, Symbol.XamarinEMDK, SkiaSharp, Rg.Plugins.Popup, ZXing.Net.Mobile.Forms, Microsoft.Data.Sqlite, MathParser.org-mXparser
- Affected Devices: All

**Comment by jsuarezruiz** [comment]
@dmsi-leo Would it be possible to have an example where test and reproduce the issue?

**Comment by dmsi-leo** [comment]
@jsuarezruiz Thank you for quick response.

I can't think of easy way to reproduce the issue.

We have our ASMX web service and backend server application running on our cloud server, but you would need our XM app been properly setup and call several web methods in specific order to generate multiple test files to download. Also we would need to set up temporary security accounts to enable web service access.

Our XM app uses on platform (Android/iOS) generated web service proxy classes to connect to the web service. We were able to trace the bottleneck to the calls synchronously invoking web proxy functions. I could send you the source code for the proxy and snippets of the code that invoke proxy functions if this can help to determine the cause of the problem.

Another option could be to enable some sort of screen sharing so you can see source code and run app in debug mode. 

Thanks for help,

Leo

**Comment by dmsi-leo** [comment]
Thank you for quick response.
I can't think of easy way to reproduce the issue.
We have our ASMX web service and backend server application running on our cloud server, but you would need our XM app been properly setup and call several web methods in specific order to generate multiple test files to download. Also we would need to set up temporary security accounts to enable web service access.
Our XM app uses on platform (Android/iOS) generated web service proxy classes to connect to the web service. We were able to trace the bottleneck to the calls synchronously invoking web proxy functions. I could send you the source code for the proxy and snippets of the code that invoke proxy functions if this can help to determine the cause of the problem.
Another option could be to enable some sort of screen sharing so you can see source code and run app in debug mode.
Thanks for help,
Leo

**Comment by hartez** [comment]
> Is it possible that performance can be improved by re-generating proxy classes with VS 2019?

@dmsi-leo It might be worth a try. 

> We were able to trace the bottleneck to the calls synchronously invoking web proxy functions.

Are you saying that with VS 2017 and Forms 4.2, the functions were called asynchronously? And that with 2019 and 4.6, they are now being called synchronously?

**Comment by dmsi-leo** [comment]
@hartez No. Proxy functions were called synchronously with VS 2017 XF 4.2 as well as with VS 2019 XF4.6. We didn't make any code changes as far as calling web service proxy functions. Also we used the same set of Android phones when tested  our app with VS 2017 and VS 2019.
Generally for performance testing we use 2 set of phones running following Android versions:
4.4 API level 19 and 7.1 API level 25.

Thanks.

**Comment by hartez** [comment]
@dmsi-leo Do you have any sort of test harness for your generated web service functions? Perhaps a console application which can call those methods, independent of your Forms application? That would let us know if the issue is specifically something in Forms, or something to do with .NET or Visual Studio.

Also, have you tried building this application on VS 2019 with XF 4.2? If the performance issues persist with that combination, we could rule out Forms as the culprit (or vice versa).

**Comment by dmsi-leo** [comment]
@hartez Sorry it took me 3 days to get back to you but it was weekend and also we needed to rebuild couple of other WPF apps that use the same web service with VS 2019.

1) So as I mentioned we have 2 WPF apps that connect to the same web service and use generated proxy classes for call web service.
After comparing VS 2017 and VS 2019 builds we didn't notice any performance difference and file download speed is identical.

2) We deleted and regenerated new web service proxy classes for our XF 4.6 app with VS 2019 and run performance test again. The results are still the same: performance is still 30-40% slower than VS 2017 XF 4.2 build.

Thanks

**Comment by hartez** [comment]
@dmsi-leo If you open your app in VS 2019 and downgrade Forms from 4.6 to 4.2, does the speed improve?

**Comment by dmsi-leo** [comment]
@hartez It's not that simple because we use a lot of different Nuget packages that also need to be downgraded in order to build app. I will try to think of something and get back to you as soon as I have results.

**Comment by dmsi-leo** [comment]
@hartez We build our app with VS 2019 and XF 4.2.
There were no speed improvements. Still **35-45% slow** then VS 2017 XF 4.2. Actually app built with VS 2019 XF 4.6 performs on average better than VS 2019 XF 4.2 by about **10%**.

**Comment by hartez** [comment]
@dmsi-leo Would you be able to share the project options you have selected for your Android and iOS projects?

**Comment by dmsi-leo** [comment]
@hartez Do you need only project files or any other information?

**Comment by hartez** [comment]
Project files, or even just screenshots of the project property pages for the iOS and Android projects. Mostly I'm interested in the "Android Options" tab and the "iOS Build" tab, but I'll take all the info I can get :)

**Comment by dmsi-leo** [comment]
[MInspect.Projects.zip](https://github.com/xamarin/Xamarin.Forms/files/4795304/MInspect.Projects.zip)

**Comment by hartez** [comment]
@dmsi-leo This is a long shot, but I believe the concurrent garbage collector is turned on by default in VS 2019. Could you try disabling that and running your tests? For your Android project, the option is in the project properties, in the Android Options tab - "Use the concurrent garbage collector".

That option was _not_ the default in VS 2017, so I'm wondering if that change is causing a problem for your application.

**Comment by dmsi-leo** [comment]
@hartez Concurrent garbage collector was off. I tried to turn it on and off but it didn't make any difference.

**Comment by hartez** [comment]
@dmsi-leo What does your application do with the files when they are downloaded? Are they saved locally? Is there any other processing/parsing that happens when they are downloaded? If they are saved locally, what are you using to accomplish that in your app?

**Comment by dmsi-leo** [comment]
@hartez Yes we store files locally but during our tests we removed all file saving/processing/parsing and measured performance only for downloading files as byte arrays. But even when we enabled back files processing it added only a small fraction to overall performance time. In our testing all file processing took less then 1% of file download time.

**Comment by PureWeen** [comment]
@dmsi-leo if you change the GC bridge options do you see any change?

https://docs.microsoft.com/en-us/xamarin/android/internals/garbage-collection#gc-bridge-options

Try all of them but I'm curious if "New" changes anything.

> No. Proxy functions were called synchronously with VS 2017 XF 4.2 as well as with VS 2019 XF4.6. We didn't make any code changes as far as calling web service proxy functions. 

How are you calling them synchronously? What does that code look like? Do they have async apis you can use?

Do you see the performance decrease with Release builds as well? 

If you look at the application output  or the logcat output and compare them on vs2017 to vs2019 do you see anything different?

Since you are downloading 150 files have you timed each file download to see if the download of each file is slower? 

Maybe I missed it but what are you doing with the 150 files? Do you save them locally or just in memory?


> We were able to trace the bottleneck to the calls synchronously invoking web proxy functions.

So if you only make one single call to the web proxy you can see a significant time difference?  

If you create a standalone app and generate a web proxy to a local host do you see the time difference there or only within the context of your app?

**Comment by PureWeen** [comment]
@dmsi-leo more info about what the client parts look like would be helpful. It'd be useful to see how exactly you are calling the endpoints and allocating the data after each call.

You might look into using an `ArrayPool` on the client side so you aren't allocating as much as well.

**Comment by dmsi-leo** [comment]
@PureWeen 
1. We tried GC bridge  different settings but it didn't change performance much. Using 'New' slightly improved performance but that was almost negligible - less than 3%.
2. Now we changed all web service calls to use asynchronous proxy functions (i.e. GetFileAsync()) that improved performance quite significantly ~ 25-30%. But that was expected and still doesn't resolve the original issue that synchronous web service proxy calls in VS 2019 build are slower than VS 2017 build with no client code changes.
3. Performance decreased in Release builds as well, we didn't bother to run performance tests for Debug builds. 
4. After downloading each file as byte array object from web service our app immediately saves it to the file before getting next file from web service. But the time spent on saving byte array object as a file is insignificant compare to time spent on downloading file from web service.
5. Downloading 1 file from web service in VS 2019 build is slower than VS 2017 build. But we need to download multiple files to get more accurate performance results since sometimes web service (IIS) response time can vary from call to call.
6. Using ArrayPool<byte> objects to allow large arrays pooling looks promising but it would probably take us some time to implement this approach. The client call to download file from web service is very simple:

 // Get file as byte[] from web service:
  byte[] oBuffer = proxy.DownloadFile(string strFileName);
// Save byte array buffer to the local file:
  SavetoFile(strFileName, oBuffer);

  App does not allocate any new buffers after receiving byte array from proxy. And again the time spent on saving byte array to 
  local file is insignificant (maybe less than 3% of web service call).

Thanks.

**Comment by PureWeen** [comment]
Without a sample project @dmsi-leo I don't think we'll be able to guide you on this one very well

Can you create a standalone project that just uses a local website to host files that demonstrates the problem? Or that calls out to some public api that you can download files from?

There's a ton of things that have changed from VS 2017 to VS 2019 with regards to mono/xamarin.android and it's not really possible to just guess anymore one without a demonstration of the problem so it can be directly analyzed.

**Comment by dmsi-leo** [comment]
@PureWeen
 I created test solution in VS 2017 containing web service ASMX app ( TestWebService1 ) and XF client (TestMobileClient1) projects.

TestWebService1 project includes "TestFiles" folder containing test files I was using for performance testing. To reduce solution zip file I left only 10 files in "TestFiles" folder but during our testing this folder contained 140 files to get consistent test results.

To set up web service please follow these steps in IIS Manager:
1. Create TestWebService1 application pool and enable 32 bit applications (see attached IIS_ApplicationPoolSetup1.PNG).
2. Create web application TestWebService1 and configure virtual directory "TestFiles" that use project's "TestFiles" folder location (see attached WebServiceSetup1.PNG).
3. Enable ASP.NET Impersonation and Basic Authentication (see attached WebServiceSetup2.PNG).

To test VS 2019 performance just open project in VS 2019 and let VS convert all project. The only change I made was to set Target Android version to 10.0 (API Level 29) and updated XF to 4.8.0.1269.

In our testing:
VS 2017 TestMobileClient1 build takes on average 110 seconds to download all files.
VS 2019 TestMobileClient1 build takes on average 170 seconds to download all files.

Please let me know if you need more information.

Thanks.

[TestAsmxService.zip](https://github.com/xamarin/Xamarin.Forms/files/5044431/TestAsmxService.zip)
![IIS_ApplicationPoolSetup1](https://user-images.githubusercontent.com/56408036/89698391-ca300600-d8d5-11ea-8973-6d43711d4cdd.PNG)
![WebServiceSetup](https://user-images.githubusercontent.com/56408036/89698392-cac89c80-d8d5-11ea-83ed-ccccd5cf1fb5.PNG)
![WebServiceSetup1](https://user-images.githubusercontent.com/56408036/89698393-cb613300-d8d5-11ea-9480-065c6a542e34.PNG)
![WebServiceSetup2](https://user-images.githubusercontent.com/56408036/89698394-cb613300-d8d5-11ea-9971-3392fcc5e9b8.PNG)
![IIS_ApplicationPoolSetup](https://user-images.githubusercontent.com/56408036/89698395-cb613300-d8d5-11ea-9877-51cfa4a900cc.PNG)

**Comment by dmsi-leo** [comment]
@PureWeen

Just wanted to add more details:
1. As I mentioned earlier we observed similar test results on iOS which lead me to believe it's not only Android issue.
2. Android device we used for testing is Zebra Technologies TC700H (Android 4.4 - API 19).

Thanks

**Comment by PureWeen** [comment]
@samhouts can you move this issue over to Xamarin.Android ?
This shouldn't be something that's going to be caused by Forms
